### ISSUE TITLE
Allow trait_level to sum all matching traits

### ISSUE THREAD

**Issue by saracoth** [richardwilkes/gcs#618]
In both the old Java and current Go versions of GCS, trait_level / advantage_level stops at the first matching trait and uses that level. Any matching traits beyond the first are ignored.

This causes issues when trait levels are separated. For example, when some levels of a trait are purchased with different modifiers (costs FP, pact-bound limitation, etc.). For example, alternate forms that increase existing rather than add new traits.

I think the Broad Jump attribute uses trait_level, so this can be verified by adding multiple Enhanced Move (Ground) traits to a new, blank character sheet.

Off topic: I'm happy to see your Go porting efforts went well. I didn't want to file this issue while the Java/Go versions were still being sorted out. Getting back into GCS now and pleased with all the improvements since I last used it.

**Comment by richardwilkes** [comment]
Are you referring to the expression functions here?

**Comment by saracoth** [comment]
Correct. The relevant code seems to be gcs/model/eval.go , the evalTraitLevel function.

**Comment by richardwilkes** [comment]
OK, thanks. That could certainly be done... however, it would then be at odds with the way everything else works, since a feature that gives a benefit based on level currently only knows the level of the thing it is within and has no way to group similar things together. I'm not sure if making the expression behave differently would end up being an issue or not. Typically, features give bonuses based on the level of the thing they are within. If those bonuses suddenly started applying to the merged total, I'd also have to somehow understand that the feature in multiple places should only actually be applied once. That sounds very problematic. Suggestions?

**Comment by saracoth** [comment]
Interesting. I hadn't seen any overlap between this and Features of a trait/equipment/etc. I agree having Features cooperate over multiple copies of a trait would be a mess, and I'm sorry if I gave the impression that was my request. Custom attributes as they are should work fine for traits with a multiplicative, exponential, or Size/Speed/Range-based impact. Provided there's a clean way for expressions to get the total levels of relevant traits.

Currently, these expressions are used in attributes, pools, and maybe some ancestry code. These sheet-level things aren't "within" any particular trait, that I can see. In fact, it seems that `trait_level` exists *because* attribute expressions aren't within any trait or Feature, and have no access to level information otherwise.

The current behavior of `evalTraitLevel` looks more like a side effect of code optimization. "We found *the* match, obviously there couldn't possibly be another, so let's not waste time looking." It has no favoritism for or attachment to one copy of a trait vs another. If we create a new sheet, give it "Enhanced Move (Ground) 1" and "Enhanced Move (Ground) 2", the value of Broad Jump will change depending on which one we happen to drag to the top. (Nice that the recalculation logic is solid enough to notice that, BTW!)

With the "just give me a level number, I don't care where it comes from" nature of `trait_level`, it arguably doesn't matter whether it picks the first match, the second, the last, or anything in between. Given that, I don't foresee anything objecting to `evalTraitLevel` totaling levels across all the matches instead of whichever it happens to find first. But maybe you see problems I haven't yet?

Thanks for your time. Sorry for not better explaining myself earlier.

**Comment by richardwilkes** [comment]
OK, so I think I understand and agree that having the expression function sum the levels of all it finds is reasonable. I'll make that change.
